src/actions/sale/purchaseDomains.ts

```diff
@@ -0,0 +1,166 @@
+
+  errorCheck(
+    domainsSold.gte(numberForSale),
+    "There are no domains left for purchase in the sale"
+  );
+
+  const address = await signer.getAddress();
+  const balance = await signer.getBalance();
+  const saleId = await contract.saleId();
+  const price = saleConfiguration.salePrice;
+  const privatePrice = saleConfiguration.privateSalePrice;
+  const saleStartTime = (await contract.saleStartBlockTimestamp()).toNumber();
+  const privateSaleDuration = saleConfiguration.mintlistSaleDuration.toNumber();
+
+  let accessList;
+  try {
+    // If using a Gnosis safe as the seller wallet you must
+    // provide the implementation address for the tx accessList
+    const sellerWallet = saleConfiguration.sellerWallet
+
+    const sellerContract = new ethers.Contract(
+      sellerWallet,
+      abi,
+      contract.provider
+    );
+    const implAddress = await sellerContract.masterCopy();
+    accessList = generateAccessList(address, sellerWallet, implAddress);
+  } catch (e) {
+    // We should not be attempting to use accessList at this point, as this is not supported by metamask:
+    // https://github.com/MetaMask/metamask-extension/issues/11863
+    //
+    // I would prefer to just remove this whole block at the moment, but in the interest of introducing minimal
+    // last minute changes, will leave as-is.
+    //
+    // If you are copying this for future use, DO NOT USE accessList without first verifying that support
+    // has been added to metamask!
+    // - Joel Tulloch
+    // -TODO-REQ parking lot with Joel
+  }
+
+  let tx: ethers.ContractTransaction;
+  const purchased = await contract.domainsPurchasedByAccountPerSale(saleId, address);
+  const currentBlock = await contract.provider.getBlock("latest");
+
+  if (status === SalePhase.Private && !isTransitionToPublicPhasePending(saleStartTime, privateSaleDuration, currentBlock)) {
+    let userClaim: Maybe<Claim> = mintlist.claims[address];
+
+    // To purchase in private sale a user must be on the mintlist
+    errorCheck(userClaim === undefined, "User is not part of private sale");
+    userClaim = userClaim!;
+
+    // Cannot purchase over the allowed mintlist limit
+    errorCheck(
+      purchased.add(count).gt(userClaim.quantity),
+      `This user has already purchased ${purchased.toString()} and buying ${count.toString()} more domains would go over the
+      maximum purchase amount of domains for this user, ${userClaim.quantity}. Try reducing the purchase amount.`
+    );
+    errorCheck(
+      balance.lt(privatePrice.mul(count)),
+      `Not enough funds given for purchase of ${count} domains`
+    );
+  
+
+    tx = await contract
+      .connect(signer)
+      .purchaseDomainsPrivateSale(
+        count,
+        userClaim.index,
+        userClaim.quantity,
+        userClaim.proof,
+        {
+          value: privatePrice.mul(count),
+          type: 2,
+          accessList: accessList,
+        }
+      );
```

Comment:
do we want to wait here for confirmations?

Reply:
It's a good thought. My concern is it's going to affect the responsiveness of the sale workflow from a user perspective.  

Most wallet apps also let you know when your transaction has been confirmed.  What do you think?

Reply:
yes, I agree, generally. responsiveness and security are on the opposite sides when it comes to smart contracts. I think this is fine, just wanted you to think about it and keep this in mind. Might come up later.